tools/clang/lib/AST/Type.cpp

```diff
@@ -446,6 +446,27 @@ const RecordType *Type::getAsStructureType() const {
   return nullptr;
 }
 
+const RecordType *Type::getAsStructureOrClassType() const {
```

Comment:
It looks to me like we don't actually need this, we could instead just use `getAsRecordType()` (and maybe `isStructureOrClassType()`) in all the places we're using this.

Reply:
There is no such method as `getAsRecordType()`.  If you mean `getAs<RecordType>()`, I suppose that would be fine.  I wasn't sure if it was truly equivalent, but I guess it only additionally includes unions, which haven't been enabled yet.  I'm not sure we want all these cases to allow unions when those are enabled though.  I guess I just haven't thought through each of these scenarios with the potential of unions.

It just seemed odd that there was a `getAsStructureType()` that would only work if it was declared `struct` but nothing that would work if it was `struct` or `class`, yet those two are almost semantically identical in C++ (and are semantically identical in HLSL), while unions are not.  I think just about everywhere we used `getAsStructureType()` we really wanted a RecordType for either `struct` or `class`, except for some weird built-in HLSL object detection that uses the class path to potentially drill into a template specialization.

Reply:
By the way, `getAs<RecordType>()` was my first thought as to how this should be fixed.  But then I thought maybe that would be too broad, and what we really want is either a `struct` or `class` type.

Reply:
Makes sense. My reading of the code was that all the places where this was being used are pretty well already assured to be struct or class. I suppose there might be some complications with union types, but we can look at that when we introduce them.